pepper-apis/src/main/java/org/broadinstitute/ddp/service/FileUploadService.java

```diff
@@ -0,0 +1,336 @@
+                queue = stream.collect(Collectors.toCollection(ArrayDeque::new));
+            }
+            if (queue.isEmpty()) {
+                break;
+            }
+
+            int numFound = queue.size();
+            while (!queue.isEmpty()) {
+                FileUpload upload = queue.remove();
+                String bucket = getBucketForUpload(upload);
+                Blob blob = storageClient.getBlob(bucket, upload.getBlobName());
+                if (blob != null && blob.exists()) {
+                    try {
+                        storageClient.deleteBlob(blob);
+                        LOG.info("Deleted blob {} for unused file upload {}", blob.getBlobId(), upload.getGuid());
+                    } catch (Exception e) {
+                        LOG.error("Unable to delete file {} from bucket {}, skipping removal of unused file upload {}",
+                                upload.getBlobName(), bucket, upload.getGuid());
+                        continue;
+                    }
+                } else if (upload.getUploadedAt() != null) {
+                    LOG.error("Unable to locate file {} in bucket {}, skipping removal of unused file upload {}",
+                            upload.getBlobName(), bucket, upload.getGuid());
+                    continue;
+                } else {
+                    // Nothing to do. Actual file is not uploaded to bucket.
+                }
+                uploadIdsToDelete.add(upload.getId());
+            }
+
+            offset += numFound;
+        }
+
+        uploadDao.deleteByIds(uploadIdsToDelete);
+        LOG.info("Removed {} unused file uploads", uploadIdsToDelete.size());
+        return uploadIdsToDelete.size();
+    }
+
+    public enum VerifyResult {
+        /**
+         * File hasn't been uploaded yet.
+         */
+        NOT_UPLOADED,
+        /**
+         * File does not belong to participant or the study.
+         */
+        OWNER_MISMATCH,
+        /**
+         * File was scanned already and found to be infected.
+         */
+        QUARANTINED,
+        /**
+         * File size does not match up.
+         */
+        SIZE_MISMATCH,
+        /**
+         * No issues found.
+         */
+        OK,
+    }
+
+    public static class AuthorizeResult {
+        private boolean exceededSize;
+        private FileUpload fileUpload;
+        private URL signedUrl;
+
+        public AuthorizeResult(boolean exceededSize, FileUpload fileUpload, URL signedUrl) {
+            this.exceededSize = exceededSize;
+            this.fileUpload = fileUpload;
+            this.signedUrl = signedUrl;
+        }
+
+        public boolean isExceededSize() {
+            return exceededSize;
+        }
+
+        public FileUpload getFileUpload() {
+            return fileUpload;
+        }
+
+        public URL getSignedUrl() {
+            return signedUrl;
+        }
+    }
+}
```

Comment:
When I was asking about shared service code with A-T, this is the sort of thing I had in mind. Perhaps they way they ended up implementing things not amenable for sharing?

Reply:
Yes, as far as I know, A-T went a different route.